Docs for Gloo Fed canary upgrade (#7516)
solo-io/gloo: docs/content/operations/upgrading/canary.md

3. If you have Gloo Edge Enterprise or Federation, set your license key as an environment variable. To request a license, [contact Sales](https://www.solo.io/company/contact/).
   ```
   export GLOO_LICENSE=<license>
   ```
4. Set the target version that you want to upgrade to as an environment variable. To find available versions, check the [changelog]({{< versioned_link_path fromRoot="/reference/changelog/" >}}). The following commands include the latest versions for each of the following Gloo Edge installation options.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
export TARGET_VERSION={{< readfile file="static/content/version_geoss_latest.md" markdown="true">}}
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell" >}}
export TARGET_VERSION={{< readfile file="static/content/version_gee_latest.md" markdown="true">}}
{{< /tab >}} 
{{< tab name="Federation" codelang="shell" >}}
export TARGET_VERSION={{< readfile file="static/content/version_gee_latest.md" markdown="true">}}
{{< /tab >}} 
   {{< /tabs >}}
5. Upgrade your `glooctl` CLI to the version that you want to upgrade to.
   ```

>    glooctl upgrade --release v${TARGET_VERSION}

Reviewer: for EE/Fed, the glooctl version should actually be the OSS version corresponding to the TARGET_VERSION. we can link to https://docs.solo.io/gloo-edge/latest/operations/upgrading/upgrade_steps/#step-2-upgrade-glooctl where it explains how to find the corresponding oss version

   ```shell
   gloooctl uninstall -n gloo-system
   ```

### In-place canary upgrades by using xDS relay {#canary-xds-relay}

By default, your Gloo Edge or Gloo Edge Enterprise control plane and data plane are installed together. However, you can decouple the control plane and data plane lifecycles by using the [`xds-relay`]({{< versioned_link_path fromRoot="/operations/advanced/xds_relay/" >}})
project as the "control plane" for the newer version deployment in a canary upgrade. This setup provides extra resiliency for your live xDS configuration in the event of failure during an in-place `helm upgrade`. 

## Upgrade Gloo Edge Federation in a canary model {#canary-upgrade-fed}

You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a management cluster that runs Gloo Edge Federation (and optionally, other Enterprise components). Then you register remote clusters with Gloo Edge Enterprise.

> In the canary upgrade model, you start with an older version of Gloo Edge Federation in your management cluster, managing one or more remote Gloo Edge instances. To test a newer target version, you install the target version of Gloo Edge Federation in a new namespace on the management cluster. Then, you install one or more Gloo Edge instances at the target version in remote test clusters, register the new clusters, and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters and uninstall the older Gloo Edge Federation version.

Reviewer 2: ```suggestion
In the canary upgrade model, you start with an older version of Gloo Edge Federation in your management cluster that manages one or more remote Gloo Edge instances. To test a newer target version, you install the target version of Gloo Edge Federation in a new namespace on the management cluster. Then, you install one or more Gloo Edge instances at the target version in remote test clusters, register the new clusters, and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters and uninstall the older Gloo Edge Federation version.
```

Reviewer 2: `Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters and uninstall the older Gloo Edge Federation version.`

Is the second piece `register the remote clusters` correct here?

Author: No, it was leftover. Will revise.

1. Update and pull the Gloo Edge Helm chart for the target version.
   ```shell
   helm repo add gloo-fed https://storage.googleapis.com/gloo-fed-helm
   helm repo update
   helm pull gloo-fed/gloo-fed --version $TARGET_VERSION --untar
   ```
2. Apply the new Gloo CRDs from the target version Helm chart to your management cluster.
   ```
   kubectl apply -f gloo-fed/charts/gloo/crds
   ```
3. Install the target version of Gloo Edge Federation in the new namespace in your management cluster.
   ```
   helm install -n gloo-system-$TARGET_VERSION gloo-fed gloo-fed/gloo-fed --create-namespace --set-string license_key=$GLOO_LICENSE --version $TARGET_VERSION
   ```
4. Verify that your current and target versions of Gloo Edge Federation are running. 
   ```shell
   kubectl get all -n gloo-system-$OLD_VERSION
   kubectl get all -n gloo-system-$TARGET_VERSION
   ```
5. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) on one of your remote clusters. Make sure that the version of Gloo Edge Enterprise matches the version of Gloo Edge Federation that you installed earlier.

> 6. In your management cluster, [create any federated custom resources]({{< versioned_link_path fromRoot="/guides/gloo_federation/federated_configuration/" >}}) for the target version in your `gloo-system-$TARGET_VERSION` namespace. You might use the existing federated resources in the `gloo-system-$OLD_VERSION` namespace as a starting point. Include any changes or new capabilities that you noticed in the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version. 

Reviewer: since each gloo fed instance will process all the federated resources on the cluster, the federated resources don't need to be created in any particular namespace. e.g. our example below shows that they can live in a shared namespace. i think we can say something like this, feel free to change the wording:

```suggestion
6. In your management cluster, [create or modify any federated custom resources]({{< versioned_link_path fromRoot="/guides/gloo_federation/federated_configuration/" >}}) needed to test your target version of Gloo Federation/Edge. Since Gloo Edge Federation will scan all federated resources on the management cluster, you may wish to reuse the existing federated resources in the `gloo-system-$OLD_VERSION` namespace. Include any changes or new capabilities that you noticed in the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version.
```

>    {{% notice note %}}**Example**: The following configuration creates a federated virtual service with different features across versions 1.12 and 1.13{{% /notice %}}

Reviewer 2: ```suggestion
   {{% notice note %}}**Example**: The following configuration creates a federated virtual service with different features across versions 1.12 and 1.13.{{% /notice %}}
```

         - gloo-system
     template:
       metadata:
         name: my-vs
       spec:
         virtualHost:
           domains:
           - '*'
           routes:
           - matchers:
             - prefix: /
             routeAction:
               single:
                 upstream:
                   name: default-petstore-8080
                   namespace: gloo-system
             options:
               retries:
                 retryOn: 'connect-failure'
                 numRetries: 10
                 perTryTimeout: '5s'
                 retryBackOff:
                   baseInterval: 1s
                   maxInterval: 3s
   EOF
   ```
1. Check that the updated, federated custom resources from the management cluster are propagated to the remote test cluster.
   1. To find the clusters and names of the propagated resources, review the federated resource configuration. The following example is based off the federated virtual service that you created in the previous step. The `PLACED` state in the status section confirms that the resource is federated.
      ```sh
      kubectl get federatedvirtualservice -n gloo-fed -o yaml
      ```
      Example truncated output:

>       {{< highlight yaml "hl_lines=9" >}} 

Reviewer: ```suggestion
      {{< highlight yaml "hl_lines=10" >}} 
```

...

> status:
>   placementStatus:
>     clusters:
>       remote1:
>         namespaces:
>           gloo-system:
>             state: PLACED

Reviewer: ```suggestion
status:
  namespacedPlacementStatuses:
    gloo-system-$TARGET_VERSION:
      clusters:
        remote2:
          namespaces:
            gloo-system:
              state: PLACED
```
the placement statuses are now namespaced by the gloo-fed namespace, so the status would look something like this (and there would also be an entry for gloo-system-<old version>)

      {{< highlight yaml "hl_lines=5" >}}
kubectl get virtualservices -n gloo-system --context remote1 -o yaml
...
options:
  retries:
    numRetries: 10
    perTryTimeout: 5s
    retryOn: connect-failure
      {{< /highlight >}}
      * In the target version's remote cluster, verify that the resource is federated with only the new version's configuration. In the example, the new 1.13 version has an updated `numRetries: 10`, as well as a `retryBackOff` section.
      {{< highlight yaml "hl_lines=6 9-11" >}} 
kubectl get virtualservices -n gloo-system --context remote2 -o yaml
...
options:
  retries:
    numRetries: 10
    perTryTimeout: 5s
    retryOn: connect-failure
    retryBackOff:
      baseInterval: 1s
      maxInterval: 3s
      {{< /highlight >}}
   3. Repeat the previous steps for each federated resource that you want to check.
8. Test your routes and monitor the metrics of the target version in your remote test cluster.
    ```shell
    glooctl check
    ```
9. Shift traffic to the target version of Gloo Edge Federation.

>    1. [Deregister]({{< versioned_link_path fromRoot="/reference/cli/glooctl_cluster_deregister/" >}}) your other remote clusters that still use the old version of Gloo Edge Federation.

Reviewer: i think 'deregister' should be bullet 10. instead of a subsection of 9?

>    2. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) on each of your remote clusters. Make sure that the version of Gloo Edge Enterprise matches the version of Gloo Edge Federation that you installed earlier.

Reviewer: this steps seems to be repeated from above, we can remove it

   3. Optionally delete the old version namespace from each remote cluster. The deregister command does not clean up the namespace and custom resources in the old version.
      ```shell
      kubectl delete ns gloo-system-$OLD_VERSION
      ```

> 10. [Uninstall]({{< versioned_link_path fromRoot="/reference/cli/glooctl_uninstall/" >}}) the older version of Gloo Edge Federation so that your management cluster uses the newer version going forward.

Reviewer: oh i see. i think the 1 and 3 sub-bullets above should all go in this section here about uninstalling the old version

Author: ok yeah I see how that was confusing. I meant it more like the way to shift the traffic to the new target version is by uninstalling the old one so that only the new target version is left. I will combine these steps.

## Upgrade Gloo Edge Federation in a canary model {#canary-upgrade-fed}

You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a management cluster that runs Gloo Edge Federation (and optionally, other Enterprise components). Then you register remote clusters with Gloo Edge Enterprise.

In the canary upgrade model, you start with an older version of Gloo Edge Federation in your management cluster that manages one or more remote Gloo Edge instances. To test a newer target version, you install the target version of Gloo Edge Federation in a new namespace on the management cluster. Then, you install one or more Gloo Edge instances at the target version in remote test clusters, register the new clusters, and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation and uninstall the older Gloo Edge Federation version from the management cluster.

> 1. Update and pull the Gloo Edge Helm chart for the target version.

Reviewer: ```suggestion
1. Update and pull the Gloo Fed Helm chart for the target version.
```

   ```shell
   helm repo add gloo-fed https://storage.googleapis.com/gloo-fed-helm
   helm repo update
   helm pull gloo-fed/gloo-fed --version $TARGET_VERSION --untar
   ```

> 2. Apply the new Gloo CRDs from the target version Helm chart to your management cluster.

Reviewer: ```suggestion
2. Apply the new Gloo Fed CRDs from the target version Helm chart to your management cluster.
```

   ```

>    kubectl apply -f gloo-fed/charts/gloo/crds

Reviewer: ```suggestion
   kubectl apply -f gloo-fed/crds
```

   2. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) for the rollback version's Gloo Edge Federation on each remote cluster.
   3. Optionally delete the newer target version namespace. The deregister command does not clean up the namespace and custom resources in the old version.
      ```shell
      kubectl delete ns gloo-system-$TARGET_VERSION
      ```
3. [Uninstall]({{< versioned_link_path fromRoot="/reference/cli/glooctl_uninstall/" >}}) the newer target version of Gloo Edge from your cluster.
    ```shell
    gloooctl uninstall -n gloo-system-$TARGET_VERSION
    ```
4.  Apply the Gloo CRDs from the rollback version Helm chart to your cluster. In Gloo Edge Federation, this cluster is the local management cluster.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
kubectl apply -f gloo/crds
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
kubectl apply -f gloo-ee/charts/gloo/crds
{{< /tab >}}
{{< tab name="Federation" codelang="shell">}}

> kubectl apply -f gloo-fed/charts/gloo/crds

Reviewer: ```suggestion
kubectl apply -f gloo-fed/crds
```

Now you're ready to upgrade your Gloo Edge installation. The steps vary depending on if you use [Open Source or Enterprise](#canary-upgrade), or [Federation](#canary-upgrade-fed).

## Upgrade Gloo Edge Open Source or Enterprise in a canary model {#canary-upgrade}

1. Update and pull the Gloo Edge Helm chart for the target version.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
helm repo add gloo https://storage.googleapis.com/solo-public-helm
helm repo update
helm pull gloo/gloo --version $TARGET_VERSION --untar
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
helm repo add glooe https://storage.googleapis.com/gloo-ee-helm
helm repo update
helm pull glooe/gloo-ee --version $TARGET_VERSION --untar
{{< /tab >}}
   {{< /tabs >}}
1. Apply the new Gloo CRDs from the target version Helm chart to your cluster.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
kubectl apply -f gloo/crds
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
kubectl apply -f gloo-ee/charts/gloo/crds
{{< /tab >}}
   {{< /tabs >}}

> 3. Install the target version of Gloo Edge in the new namespace in your cluster.

Reviewer 2: should the CRDs be placed in the same namespace where you install gloo edge?

Author: I'm not sure, that might mean it has to be moved after the install command since we rely on that to create the new namespace.

   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
glooctl install gateway --version $TARGET_VERSION -n gloo-system-$TARGET_VERSION
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
glooctl install gateway enterprise --version $TARGET_VERSION -n gloo-system-$TARGET_VERSION --license-key $GLOO_LICENSE
{{< /tab >}}
   {{< /tabs >}}
4. Verify that your current and target versions of Gloo Edge are running.
   ```shell
   kubectl get all -n gloo-system
   kubectl get all -n gloo-system-$TARGET_VERSION
   ```

> 5. Modify any custom resources for any changes or new capabilities that you noticed the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version.

Reviewer 2: ```suggestion
5. Modify any custom resources for any changes or new capabilities that you noticed in the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version.
```

6. Test your routes and monitor the metrics of the newer version.
    ```shell
    glooctl check
    ```
7. [Uninstall]({{< versioned_link_path fromRoot="/reference/cli/glooctl_uninstall/" >}}) the older version of Gloo Edge so that your cluster uses the newer version going forward.
   ```shell
   gloooctl uninstall -n gloo-system
   ```

## Upgrade Gloo Edge Federation in a canary model {#canary-upgrade-fed}

> You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a local cluster that runs both Gloo Edge Federation and Enterprise components. Then you register remote clusters with Gloo Edge Enterprise. In the canary upgrade model, you deploy both versions of Gloo Edge Federation to your local cluster. Then, you register a remote test cluster for the newer target version of Gloo Edge Federation and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters for the newer target version, and uninstall the older version.

Reviewer 2: ```suggestion
You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a local cluster that runs both Gloo Edge Federation and Enterprise components. Then, you register remote clusters with Gloo Edge Enterprise. In the canary upgrade model, you deploy both versions of Gloo Edge Federation to your local cluster. Then, you register a remote test cluster for the newer target version of Gloo Edge Federation and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters for the newer target version, and uninstall the older version.
```

Reviewer 2: I am wondering if you can remove the first two sentences. reading through it, it almost sounded as it was a copy paste error as the first 2 sound very similar to the 2 after that.

> You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a local cluster that runs both Gloo Edge Federation and Enterprise components. Then you register remote clusters with Gloo Edge Enterprise. In the canary upgrade model, you deploy both versions of Gloo Edge Federation to your local cluster. Then, you register a remote test cluster for the newer target version of Gloo Edge Federation and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters for the newer target version, and uninstall the older version.

Reviewer: ```suggestion
You can upgrade Gloo Edge Federation in a canary model in version 1.13 or later. In [Gloo Edge Federation]({{< versioned_link_path fromRoot="/guides/gloo_federation/" >}}), you have a management cluster that runs Gloo Edge Federation (and optionally other Enterprise components), and register remote Gloo Edge clusters to manage. In the canary upgrade model, you initially have Gloo Edge Federation deployed on your management cluster, managing one or more remote Gloo Edge instances. To test a new version of Gloo Edge Federation, you install the new Gloo Edge Federation on the management cluster, in a different namespace, install one or more Gloo Edge instances (running the new Gloo Edge version) on remote clusters, register the new clusters, and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation and uninstall the older Gloo Edge Federation version.
```

Reviewer: also, i think it may be more clear if we refer to the Gloo Fed cluster as the "management" cluster throughout (instead of "local" cluster)

1. Update and pull the Gloo Edge Helm chart for the target version.
   ```shell
   helm repo add gloo-fed https://storage.googleapis.com/gloo-fed-helm
   helm repo update
   helm pull gloo-fed/gloo-fed --version $TARGET_VERSION --untar
   ```

> 2. Apply the new Gloo CRDs from the target version Helm chart to your cluster. In Gloo Edge Federation, this cluster is the local management cluster.

Reviewer 2: same question here, do these need to go into the new namespace?

Author: see previous thread

   ```
   kubectl apply -f gloo-fed/charts/gloo/crds
   ```

> 3. Install the target version of Gloo Edge in the new namespace in your cluster.

Reviewer 2: maybe mention that you still do this in the local cluster

> 3. Install the target version of Gloo Edge Federation in the new namespace in your local cluster.

Reviewer: ```suggestion
3. Install the target version of Gloo Edge Federation in the new namespace in your management cluster.
```

   ```

>    glooctl install gateway enterprise --version $TARGET_VERSION -n gloo-system-$TARGET_VERSION --license-key $GLOO_LICENSE

Reviewer: i think we should use helm here, as `glooctl install` will install GlooEE as well as Gloo Fed. something like:
```suggestion
   helm install -n gloo-system-$TARGET_VERSION gloo-fed gloo-fed/gloo-fed --create-namespace --set-string license_key=$GLOO_LICENSE --version $TARGET_VERSION
```

   ```

> 4. Verify that your current and target versions of Gloo Edge are running. 

Reviewer 2: ```suggestion
4. Verify that your current and target versions of Gloo Edge Federation are running. 
```

   ```shell
   kubectl get all -n gloo-system-$OLD_VERSION
   kubectl get all -n gloo-system-$TARGET_VERSION
   ```

> 5. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) for the target version's Gloo Edge Federation on a remote test cluster.

Reviewer 2: ```suggestion
5. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) on a remote test cluster. Make sure that the version of Gloo Edge Enterprise matches the version of Gloo Edge Federation that you installed ealier.
```

> 6. In your remote test cluster, modify any custom resources for any changes or new capabilities that you noticed in the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version.

Reviewer: instead of modifying resources directly on the remote clusters, they should modify Gloo Fed resources on the management cluster, and make sure the changes propagate to the remote clusters as expected, and that traffic is handled as expected on the remote clusters.

7. Test your routes and monitor the metrics of the target version in your remote test cluster.
    ```shell
    glooctl check
    ```
8. Shift traffic to the target version of Gloo Edge Federation.

>    1. [Deregister]({{< versioned_link_path fromRoot="/reference/cli/glooctl_cluster_deregister/" >}}) your other remote clusters that still use the old version of Gloo Edge Federation.

Reviewer 2: is there a way to check which ones these are?

Reviewer 2: I think you can ignore this comment. I did not quite get the upgrade process

>    2. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) for the target version's Gloo Edge Federation on each remote cluster.

Reviewer 2: ignore that, but maybe take a look at the phrasing suggestion earlier

>    2. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) for the target version's Gloo Edge Federation on each remote cluster.

Author: ```suggestion
   2. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) on each of your remote clusters. Make sure that the version of Gloo Edge Enterprise matches the version of Gloo Edge Federation that you installed earlier.
```

   3. Optionally delete the old version namespace. The deregister command does not clean up the namespace and custom resources in the old version.
      ```shell
      kubectl delete ns gloo-system-$OLD_VERSION
      ```
9.  [Uninstall]({{< versioned_link_path fromRoot="/reference/cli/glooctl_uninstall/" >}}) the older version of Gloo Edge Federation so that your local cluster uses the newer version going forward.
   ```shell

>    gloooctl uninstall -n gloo-system-$OLD_VERSION

Author: oh that was in the remote cluster, whereas this is in the local cluster.

## Upgrade Gloo Edge Open Source or Enterprise in a canary model {#canary-upgrade}

1. Update and pull the Gloo Edge Helm chart for the target version.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
helm repo add gloo https://storage.googleapis.com/solo-public-helm
helm repo update
helm pull gloo/gloo --version $TARGET_VERSION --untar
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
helm repo add glooe https://storage.googleapis.com/gloo-ee-helm
helm repo update
helm pull glooe/gloo-ee --version $TARGET_VERSION --untar
{{< /tab >}}
   {{< /tabs >}}
1. Apply the new Gloo CRDs from the target version Helm chart to your cluster.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}
kubectl apply -f gloo/crds
{{< /tab >}}
{{< tab name="Enterprise" codelang="shell">}}
kubectl apply -f gloo-ee/charts/gloo/crds
{{< /tab >}}
   {{< /tabs >}}
3. Install the target version of Gloo Edge in the new namespace in your cluster.
   {{< tabs >}} 
{{< tab name="Open Source" codelang="shell" >}}

> glooctl install gateway --version $TARGET_VERSION -n gloo-system-$TARGET_VERSION

Reviewer 2: they might need to create the namespace first.

Author: the command does this for them

In the canary upgrade model, you start with an older version of Gloo Edge Federation in your management cluster, managing one or more remote Gloo Edge instances. To test a newer target version, you install the target version of Gloo Edge Federation in a new namespace on the management cluster. Then, you install one or more Gloo Edge instances at the target version in remote test clusters, register the new clusters, and check your setup. Finally, you deregister the remote clusters from the older version of Gloo Edge Federation, register the remote clusters and uninstall the older Gloo Edge Federation version.

1. Update and pull the Gloo Edge Helm chart for the target version.
   ```shell
   helm repo add gloo-fed https://storage.googleapis.com/gloo-fed-helm
   helm repo update
   helm pull gloo-fed/gloo-fed --version $TARGET_VERSION --untar
   ```
2. Apply the new Gloo CRDs from the target version Helm chart to your cluster. In Gloo Edge Federation, this cluster is the local management cluster.
   ```
   kubectl apply -f gloo-fed/charts/gloo/crds
   ```
3. Install the target version of Gloo Edge Federation in the new namespace in your management cluster.
   ```

>    helm install -n gloo-system-$TARGET_VERSION gloo-fed gloo-fed/gloo-fed --create-namespace --set-string license_key=$GLOO_LICENSE --version $TARGET_VERSION

Reviewer 2: will this command automatically create the namespace as `gloo-system-$TARGET_VERSION`?

Author: yes, because it has the `--create-namespace` flag

   ```
4. Verify that your current and target versions of Gloo Edge Federation are running. 
   ```shell
   kubectl get all -n gloo-system-$OLD_VERSION
   kubectl get all -n gloo-system-$TARGET_VERSION
   ```
5. [Install and register Gloo Edge Enterprise]({{< versioned_link_path fromRoot="/guides/gloo_federation/cluster_registration/" >}}) on one of your remote clusters. Make sure that the version of Gloo Edge Enterprise matches the version of Gloo Edge Federation that you installed earlier.

> 6. In your management cluster, modify any federated custom resources for any changes or new capabilities that you noticed in the [upgrade notice]({{< versioned_link_path fromRoot="/operations/upgrading/" >}}) and the [changelogs]({{< versioned_link_path fromRoot="/reference/changelog/" >}}) for the target version.

Reviewer 2: can we add a command for this?

Author: I'll add a link to the guide about creating fed resources

> 7. Check that the updated, federated custom resources from the management cluster are propagated to the remote test cluster. To find the clusters and names of the propagated resources, review the federated resource configuration.

Reviewer 2: can we add a command for this?

Author: adding substeps
